membership-attribute-service/app/controllers/SalesforceHookController.scala

```diff
@@ -40,99 +40,153 @@ class SalesforceHookController extends LazyLogging {
+      (for {
+        sfId <- OptionT(touchpoint.contactRepo.get(membershipUpdate.UserId).logWith("contact id from SF", _.map(_.salesforceContactId)))
+        membershipSubscription <- OptionT(touchpoint.subService.current[SubscriptionPlan.Member](sfId).logWith("current subscriptions", _.map(_.id)).map(_.headOption))
       } yield {
-        (stripeCustomer.card.exp_month, stripeCustomer.card.exp_year)
-      }).run
 
-      cardExpiryFromStripeF.map {
-        case Some((expMonth, expYear)) => salesforceAttributes.copy(Tier = tierFromZuora, CardExpirationMonth = Some(expMonth), CardExpirationYear = Some(expYear))
-        case None => salesforceAttributes.copy(Tier = tierFromZuora)
+        // If the tier info does not match, we trust the info we get from Zuora, instead of the tier sent to us in the outbound message from Salesforce
+        val tierFromZuora = membershipSubscription.plan.charges.benefit.id
+        val tierFromSalesforce = membershipUpdate.Tier
+        if (tierFromZuora != tierFromSalesforce) error(s"Differing tier info for $sfId : sf=$tierFromSalesforce zuora=$tierFromZuora")
+
+        // If we have the card expiry date in Stripe, add them to Dynamo too.
+        // TODO - refactor to use touchpoint.paymentService - requires membership-common model tweak first.
+        val cardExpiryFromStripeF = (for {
+          account <- OptionT(touchpoint.zuoraService.getAccount(membershipSubscription.accountId).map(Option(_)))
+          paymentMethodId <- OptionT(Future.successful(account.defaultPaymentMethodId))
+          paymentMethod <- OptionT(touchpoint.zuoraService.getPaymentMethod(paymentMethodId).map(Option(_)))
+          customerToken <- OptionT(Future.successful(paymentMethod.secondTokenId))
+          stripeCustomer <- OptionT(touchpoint.stripeService.Customer.read(customerToken).map(Option(_)))
+        } yield {
+          (stripeCustomer.card.exp_month, stripeCustomer.card.exp_year)
+        }).run
+
+        val startDate = membershipSubscription.startDate // acceptanceDate is the date of first payment, but we want to know the signup date - contract effective date
+
+        cardExpiryFromStripeF.map {
+          case Some((expMonth, expYear)) =>
+            Attributes(
+              UserId = membershipUpdate.UserId,
+              Tier = tierFromZuora,
+              MembershipNumber = membershipUpdate.MembershipNumber,
+              AdFree = None,
+              CardExpirationMonth = Some(expMonth),
+              CardExpirationYear = Some(expYear),
+              StartDate = Some(startDate)
+            )
+          case None =>
+            Attributes(
+              UserId = membershipUpdate.UserId,
+              Tier = tierFromZuora,
+              MembershipNumber = membershipUpdate.MembershipNumber,
+              AdFree = None,
+              CardExpirationMonth = None,
+              CardExpirationYear = None,
+              StartDate = Some(startDate)
+            )
```

Comment:
isn't there a nice way to turn this Option(expMonth,expYear) into Option(expmonth), Option(exyear) and eliminate the duplication ?